Source/DafnyCore/Triggers/TriggersCollector.cs

```diff
@@ -390,6 +390,37 @@ internal class TriggersCollector {
       return new TriggerAnnotation(children_contain_killers || expr_is_killer, new_term.Variables, collected_terms);
     }
 
+    private IEnumerable<TriggerTerm> CollectInSetOperations(Expression new_expr, Expression original_expr) {
+      yield return new TriggerTerm { Expr = new_expr, OriginalExpr = original_expr, Variables = CollectVariables(original_expr) };
+      if (new_expr is BinaryExpr
+        {
+          ResolvedOp: BinaryExpr.ResolvedOpcode.InSet,
+          E0: var e0, E1: BinaryExpr
+          {
+            ResolvedOp: BinaryExpr.ResolvedOpcode.SetDifference or
+                          BinaryExpr.ResolvedOpcode.Union or
+                          BinaryExpr.ResolvedOpcode.Intersection,
+            E0: var e10,
+            E1: var e11
+          }
+        }) {
+        var newTriggerTerm1 = new BinaryExpr(new_expr.tok, BinaryExpr.Opcode.In, e0, e10) {
+          ResolvedOp = BinaryExpr.ResolvedOpcode.InSet,
+          Type = Type.Bool
+        };
+        foreach (var triggerTerm in CollectInSetOperations(newTriggerTerm1, original_expr)) {
+          yield return triggerTerm;
+        }
+        var newTriggerTerm2 = new BinaryExpr(new_expr.tok, BinaryExpr.Opcode.In, e0, e11) {
+          ResolvedOp = BinaryExpr.ResolvedOpcode.InSet,
+          Type = Type.Bool
+        };
+        foreach (var triggerTerm in CollectInSetOperations(newTriggerTerm2, original_expr)) {
+          yield return triggerTerm;
+        }
```

Comment:
What happens for some input like

``` dafny
function F(x: int): set<int>

method Test(x: int)
  requires forall x: int, y: int, z: int :: x in F(y) * F(z)
{
}
```

Previously, this would generate `{:trigger x in F(y) * F(z)}`. Will this now additionally generate

    {:trigger x in F(y)}
    {:trigger x in F(z)}

If so, then note that these are not valid triggers, since neither one by itself covers all bound variables.

Reply:
Indeed, it now generates these additional triggers. Let me have a look.

Reply:
> If so, then note that these are not valid triggers, since neither one by itself covers all bound variables.

Fixed by checking that the number of bound variables remains the same. Good catch !